packages/wonder-blocks-modal/components/modal-launcher/index.test.js

```diff
@@ -0,0 +1,60 @@
+// @flow
+import React from "react";
+import {shallow} from "enzyme";
+
+import ModalLauncher from "./index.js";
+
+describe("ModalLauncher", () => {
+    test("Children can launch the modal", () => {
+        const wrapper = shallow(
+            <ModalLauncher modal={<div />}>
+                {({openModal}) => <button onClick={openModal} />}
+            </ModalLauncher>,
+        );
+        expect(wrapper.find("ModalLauncherPortal")).toHaveLength(0);
+        wrapper.find("button").simulate("click");
+        expect(wrapper.find("ModalLauncherPortal")).toHaveLength(1);
+    });
+
+    test("Modal can close itself after launching", (done) => {
+        let opened = false;
+
+        // Once the modal mounts, we'll immediately self-close it on the next
+        // tick, to test the children's ability to self-close.
+        //
+        // NOTE(mdr): It would be nice to have this be, like, a close button
+        //     that closes when you click it. But that requires the button to
+        //     actually _mount_, which means we'd need to do a full DOM render
+        //     including `ModalLauncherPortal`, and that seems to be going
+        //     beyond the scope of this test. Really we just want to check that
+        //     this function receives a `closeModal` argument that works.
+        const modalFn = ({closeModal}) => {
+            expect(opened).toBe(true);
+            setImmediate(closeModal);
+            return <div />;
+        };
+
+        // Once the modal closes, we'll check that it _really_ closed, and
+        // finish the test.
+        const onClose = () => {
+            wrapper.update();
+            expect(wrapper.find("ModalLauncherPortal")).toHaveLength(0);
+            done();
+        };
+
+        // Mount the modal launcher. This shouldn't trigger any closing yet,
+        // because we shouldn't be calling the `modal` function yet.
+        const wrapper = shallow(
+            <ModalLauncher modal={modalFn} onClose={onClose}>
```

Comment:
I wonder if we should have `onClose` on the modal itself?  The reasoning, if we have it on `ModalLauncher` then users may have to keep track of which modal so that they can respond appropriately.  The only reason why I put it here is b/c the ModalLauncher also needs to call something when the user taps the backdrop to dismiss the modal.  Maybe we could just grab `this.props.modal.props.onClose` and call that function from `ModalLauncher` when someone taps the backdrop?

Reply:
Hmm nod. I think it makes sense for `ModalLauncher` to be the canonical source of closedness, since it controls the closed state, and I would find it weird for children to be aware of that.

It sounds like the case you're describing is using one shared modal launcher to conditionally render multiple different kinds of modals? In that case, I wonder if it's more appropriate to have multiple modal launchers instead.

I wonder if the modal's `onClose` prop in the spec should be renamed `onClickCloseButton`, to clarify that it's a UI-coupled event rather than signaling changes to the modal's actual open/closed state?